### PR TITLE
Drivers: add lidar driver of robosense from neolix

### PR DESCRIPTION
**Pull request by pkuwangjia**

### CODE REVIEW

File: modules/drivers/lidar/lidar_robosense/conf/lidar16_left.conf [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+time_zone: 8
+firing_data_port: 2369
+positioning_data_port: 8310
+calibration_file: "/apollo/modules/drivers/lidar/robosense/params/VLP16_calibration.yaml"
+organized: true
+max_range: 70.0
+min_range: 0.4
+target_frame_id: "novatel"
+
+#scan_channel:"/sensor/suteng16/left/SutengScan"
```
Comment (SeasoulChris) [inline modules/drivers/lidar/lidar_robosense/conf/lidar16_left.conf:16]: better delete the unused comment in this file and other files

File: modules/drivers/lidar/lidar_robosense/include/driver/driver.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ public:
+  Robosense16Driver(const apollo::drivers::suteng::SutengConfig& robo_config);
+  ~Robosense16Driver();
+
+  void init();
+  bool poll(const std::shared_ptr<apollo::drivers::suteng::SutengScan>& scan);
+  void poll_positioning_packet();
+
+ private:
+  std::shared_ptr<Input> _positioning_input;
```
Comment (SeasoulChris) [inline modules/drivers/lidar/lidar_robosense/include/driver/driver.h:77]: please reform the private number variables from `_var` to `var_`

File: modules/drivers/lidar/lidar_robosense/include/component/component_convert.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#pragma once
+#include <pcl/io/pcd_io.h>
+#include <pcl/point_types.h>
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/component/component_convert.h:19]: Please adjust header orders.

File: modules/drivers/lidar/lidar_robosense/include/component/component_driver.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    writer_ = node_->CreateWriter<apollo::drivers::suteng::SutengScan>(
+        config.scan_channel());
+
+    dvr_.reset(driver);
+    dvr_->init();
+    // spawn device poll thread
+    runing_ = true;
+    device_thread_ = std::shared_ptr<std::thread>(
+        new std::thread(std::bind(&CompRoboDriver::device_poll, this)));
+    device_thread_->detach();
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/component/component_driver.h:65]: By default, a detached thread is not joinable. I personally not very clear the life cyber of a detached thread w.r.t. the main thread.

File: modules/drivers/lidar/lidar_robosense/include/component/component_driver.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    RobosenseDriver *driver =
+        robosense::RobosenseDriverFactory::create_driver(config);
+    if (driver == nullptr) {
+      return false;
+    }
+    writer_ = node_->CreateWriter<apollo::drivers::suteng::SutengScan>(
+        config.scan_channel());
+
+    dvr_.reset(driver);
+    dvr_->init();
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/component/component_driver.h:60]: Better not use not commonly recognized abbrs.  For this case, `dvr_` can be renamed to `driver_` to improve readability.

File: modules/drivers/lidar/lidar_robosense/include/lib/pcap_input.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#pragma once
+#include <pcap.h>
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/lib/pcap_input.h:18]: header orders.

<pcap.h> should be after std headers.

File: modules/drivers/lidar/lidar_robosense/include/lib/socket_input.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#pragma once
+#include <pcap.h>
+#include <stdio.h>
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/lib/socket_input.h:19]: And we prefer c++-style `<cstdio>` to raw c header `<stdio.h>`

File: modules/drivers/lidar/lidar_robosense/include/parser/convert.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#pragma once
+#include <memory>
+
+#include "modules/drivers/lidar/lidar_robosense/include/parser/robosense_parser.h"
+#include "modules/drivers/lidar/lidar_robosense/proto/sensor_suteng.pb.h"
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/parser/convert.h:21]: header orders.

File: modules/drivers/lidar/lidar_robosense/include/parser/robosense_parser.h [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#pragma once
+#include <errno.h>
+#include <stdint.h>
+
+#include <boost/format.hpp>
+#include <map>
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/parser/robosense_parser.h:22]: header orders.

File: modules/drivers/lidar/lidar_robosense/include/parser/robosense_status.h [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+#pragma once
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "cyber/cyber.h"
+#include "modules/drivers/lidar/lidar_robosense/proto/sensor_suteng.pb.h"
+#include "modules/drivers/lidar/lidar_robosense/proto/sensor_suteng_conf.pb.h"
+#include "modules/drivers/proto/pointcloud.pb.h"
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/parser/robosense_status.h:26]: header orders. 

pb files should be placed in its own group and before `cyber.h`, right @changsh726 ?

File: modules/drivers/lidar/lidar_robosense/include/parser/robosense_status.h [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace apollo {
+namespace drivers {
+namespace robosense {
+
+constexpr double RADIANS_TO_DEGREES = 180.0 / 3.1415926535897;
+constexpr double DEGRESS_TO_RADIANS = 3.1415926535897 / 180.0;
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/parser/robosense_status.h:33]: Check wether DEG_TO_RAD has been defined elsewhere in Apollo. Avoid hard-code PI value. Use the one from `<cmath>` instead.

File: modules/drivers/lidar/lidar_robosense/include/parser/robosense_status.h [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+#pragma pack(push, 1)  // Turn off struct padding.
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/include/parser/robosense_status.h:35]: On struct padding, prefer new style `alignas`.

File: modules/drivers/lidar/lidar_robosense/proto/BUILD [hunk 13]
```diff
@@ -0,0 +1,40 @@
+load("//tools:cpplint.bzl", "cpplint")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "suteng_proto",
+    deps = [
+        ":sensor_suteng_proto",
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/proto/BUILD:8]: Please rewrite the proto BUILD file using proto_build_generator.py when this PR was merged into master.

File: modules/drivers/lidar/lidar_robosense/src/driver/driver.cpp [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/drivers/lidar/lidar_robosense/include/driver/driver.h"
+
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/src/driver/driver.cpp:18]: Don't use `include`, `src/srcs` in subdirs in Apollo. Bazel doesn't like code organized that way.

File: modules/drivers/lidar/lidar_robosense/src/driver/driver.cpp [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+#include <time.h>
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/src/driver/driver.cpp:19]: `<ctime>`

File: modules/drivers/lidar/lidar_robosense/src/driver/driver16.cpp [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+                                  nmea_time->usec) *
+                1e3 +
+            timestamp_sec;  // ns
+        AINFO << "first POS-GPS-timestamp: [" << timestamp_nsec << "]";
+        basetime_ = timestamp_nsec;
+        AINFO << "frame_id:" << config_.frame_id()
+              << "-T(gps-time):" << nmea_time->year << "-" << nmea_time->mon
+              << "-" << nmea_time->day << "  " << nmea_time->hour << "-"
+              << nmea_time->min << "-" << nmea_time->sec;
+        AINFO << "first POS-GPS-time: [" << nmea_time->year << "/"
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/src/driver/driver16.cpp:171]: Why so many and maybe duplicate prints? Better write a convenient function to help readability.

File: modules/drivers/lidar/lidar_robosense/src/driver/driver16.cpp [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+              << nmea_time->sec << "-" << nmea_time->msec << "/"
+              << nmea_time->usec << "]";
+        start_time_ = apollo::cyber::Time().Now().ToNanosecond();
+        AINFO << "first start_time_:[" << start_time_ << "]";
+        if (rc == 0) {
+          break;  // got a full packet
+        }
+      }
+    }
+    if ( basetime_ != 0) break;  // temp
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/src/driver/driver16.cpp:183]: Code-style.

File: modules/drivers/lidar/lidar_robosense/src/lib/calibration.cpp [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/drivers/lidar/lidar_robosense/include/lib/calibration.h"
+
+#include <yaml-cpp/yaml.h>
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/src/lib/calibration.cpp:19]: Header orders.

File: modules/drivers/lidar/lidar_robosense/src/lib/calibration.cpp [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+
+namespace YAML {
+
+// The >> operator disappeared in yaml-cpp 0.5, so this function is
+// added to provide support for code written under the yaml-cpp 0.3 API.
+template <typename T>
+void operator>>(const YAML::Node& node, T& i) {
+  i = node.as<T>();
+}
+}  // namespace YAML
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/src/lib/calibration.cpp:37]: If only used in this file, use anonymous namespace.

**Review by storypku — CHANGES_REQUESTED** [review]

File: modules/drivers/lidar/lidar_robosense/include/component/component_convert.h (hunk 3 above)
Comment (pkuwangjia) [reply modules/drivers/lidar/lidar_robosense/include/component/component_convert.h:19]: adjust in what order？

File: modules/drivers/lidar/lidar_robosense/include/parser/convert.h (hunk 8 above)
Comment (pkuwangjia) [reply modules/drivers/lidar/lidar_robosense/include/parser/convert.h:21]: what wrong？plz be specific

File: modules/drivers/lidar/lidar_robosense/include/parser/robosense_parser.h (hunk 9 above)
Comment (pkuwangjia) [reply modules/drivers/lidar/lidar_robosense/include/parser/robosense_parser.h:22]: header order was arranged by clangformat ，what‘s wrong？

File: modules/drivers/lidar/lidar_robosense/src/driver/driver16.cpp (hunk 17 above)
Comment (pkuwangjia) [reply modules/drivers/lidar/lidar_robosense/src/driver/driver16.cpp:183]: how about if ( basetime_ ) break;

File: modules/drivers/lidar/lidar_robosense/include/parser/robosense_status.h (hunk 10 above)
Comment (pkuwangjia) [reply modules/drivers/lidar/lidar_robosense/include/parser/robosense_status.h:26]: other modules ’s header orders are not as you said

File: modules/drivers/lidar/lidar_robosense/lib/calibration.cpp [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+  correction.second.sin_vert_correction =
+      sinf(correction.second.vert_correction);
+  correction.second.focal_offset =
+      256 * pow(1 - correction.second.focal_distance / 13100, 2);
+  correction.second.laser_ring = 0;  // clear initially (set later)
+}
+
+void operator>>(const YAML::Node& node, Calibration& calibration) {
+  int num_lasers = 0;
+  // node[NUM_LASERS] >> num_lasers;
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/lib/calibration.cpp:92]: Please remove useless lines.

File: modules/drivers/lidar/lidar_robosense/parser/robosense_parser.cpp [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+  return temp;
+}
+float RobosenseParser::CalibIntensity(float intensity,
+                                      int calIdx,  // 1-16
+                                      int distance, float temper) {
+  int algDist;
+  int sDist;
+  int uplimitDist;
+  float realPwr;
+  float refPwr;
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/parser/robosense_parser.cpp:178]: Don't use `camelCase` for variable names.

File: modules/drivers/lidar/lidar_robosense/lib/socket_input.cpp [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+        //      <<nmea_time->hour<<"/"<< nmea_time->min<<"/"<<
+        //      nmea_time->sec<<"-" << nmea_time->msec
+        //      <<"/"<< nmea_time->usec<< "]";
+        flags = true;
+      }
+      // AINFO << "PPS-GPS-time: [" <<
+      // nmea_time->year<<"/"<<nmea_time->mon<<"/"<<nmea_time->day<<"-"
+      //        <<nmea_time->hour<<"/"<< nmea_time->min<<"/"<<
+      //        nmea_time->sec<<"-" << nmea_time->msec
+      //        <<"/"<< nmea_time->usec<< "]";
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/lib/socket_input.cpp:171]: Remove those lines commented out.

File: modules/drivers/lidar/lidar_robosense/lib/socket_input.cpp [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+      }
+    }
+
+    if ((size_t)nbytes == FIRING_DATA_PACKET_SIZE) {
+      pkt->set_data(bytes, FIRING_DATA_PACKET_SIZE);
+      // uint64_t temp_stamp_ = apollo::cyber::Time().Now().ToNanosecond();
+      // uint64_t pkt_start_diff_ = temp_stamp_ -  start_time_;
+      // AINFO<<"pkt start_time_:["<<_start_time<<"]";
+      // AINFO<<"_pkt_start_diff:"<<apollo::cyber::Time( pkt_start_diff_ -
+      // last_pkt_stamp).ToSecond()*1000<<" ms"; last_pkt_stamp =
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/lib/socket_input.cpp:119]: Remove lines commented out.

File: modules/drivers/lidar/lidar_robosense/lib/input.h [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+#include "modules/drivers/lidar/lidar_robosense/lib/data_type.h"
+#include "modules/drivers/lidar/lidar_robosense/proto/sensor_suteng.pb.h"
+
+namespace apollo {
+namespace drivers {
+namespace robosense {
+
+static const size_t FIRING_DATA_PACKET_SIZE = 1248;
+static const size_t POSITIONING_DATA_PACKET_SIZE =
+    1248;  // beike-256  robosense-512;
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/lib/input.h:32]: Have you formatted your code?

File: modules/drivers/lidar/lidar_robosense/lib/calibration.cpp [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+        min_seen = angle;
+        next_index = j;
+      }
+    }
+
+    if (next_index < num_lasers) {  // anything found in this ring?
+      // store this ring number with its corresponding laser number
+      calibration.laser_corrections_[next_index].laser_ring = ring;
+      next_angle = min_seen;
+      //        ROS_INFO_STREAM("laser_ring[" << next_index << "] = " << ring
```
Comment (storypku) [inline modules/drivers/lidar/lidar_robosense/lib/calibration.cpp:128]: Remove commented out lines.

**Review by storypku — CHANGES_REQUESTED** [review]

File: modules/drivers/lidar/lidar_robosense/lib/input.h (hunk 24 above)
Comment (pkuwangjia) [reply modules/drivers/lidar/lidar_robosense/lib/input.h:32]: yes

**Review by SeasoulChris — APPROVED** [review]

**Pull request merged**